HttpClient updates for standalone & functional interceptors

alxhub opened the pull request:
plus a new behavior for interceptors where they're additive when lazy
loaded

AndrewKushnir commented on packages/common/http/src/http.ts:
```diff
@@ -0,0 +1,177 @@
+
+  addInterceptors(fns: HttpInterceptorFn[]) {
+    for (const fn of fns) {
+      this.interceptorFns.add(fn);
```
I'm wondering if it might make sense to clear the set of `this.interceptorFns` in `ngOnDestroy` to avoid memory leaks?

AndrewKushnir commented on packages/common/http/src/http.ts:
```diff
@@ -0,0 +1,177 @@
+  constructor(private backend: HttpBackend, private injector: EnvironmentInjector) {}
+
+  private interceptorFns = new Set<HttpInterceptorFn>();
+  private haveAddedInterceptorsFromDi = false;
```
It doesn't seem like we flip this flag anywhere, may be we don't need it?

AndrewKushnir reviewed (commented):
Looks great! 👍

alxhub committed:
refactor(http): reorganize the code before new features/updates

This commit reorganizes the HttpClient codebase. In particular, it combines
a number of classes/interfaces which previously existed in separate files
into a single `http.ts` file. This is necessary in preparation for the
switch to tree-shakable providers, which will introduce new dependency
arrows that would otherwise result in cycles.

alxhub committed:
refactor(http): add internal support for functional interceptors

This commit refactors the interceptor system in HttpClient to use functional
interceptors under the hood. Existing DI-injected interceptors are converted
to functional interceptors when loaded.

This is done for two reasons:

* in support of a future commit which will expose functional interceptors
  as a new feature
* to ease the implementation of `providedIn: 'root'` support for HttpClient,
  which requires the XSRF protection interceptor to be loaded without any DI
  config). Loading it as a functional interceptor is straightfoward.

Technically, this change allows `inject()` to work within existing
interceptors, but the future introduction of functional interceptors will
serve as the "official" release of this functionality.

alxhub committed:
refactor(http): update the XSRF protection interceptor to be functional

This commit replaces the class-based XSRF interceptor with a functional one.
Instead of providing it via DI, it's always provided as a default
functional interceptor. Other XSRF-related tokens (`XSRF_COOKIE_NAME` for
example) are also converted to `providedIn: 'root'`.

Overall, this means that the default XSRF setup no longer requires including
special DI providers to operate. This is a prerequisite for making the whole
`HttpClient` available in the root injector automatically.

alxhub committed:
feat(http): introduce functional HTTP interceptors

This commit adds a new flavor of HTTP interceptor that's defined as a
function instead of a class. For example, the basic interceptor:

```typescript
@Injectable()
class NoopInterceptor implements HttpInterceptor {
  intercept(req: HttpRequest<unknown>, next: HttpHandler):
      Observable<HttpEvent<unknown>>) {
    return next.handle(req); // do nothing
  }
}
```

could be instead written as:

```typescript
function noopInterceptor(req: HttpRequest<unknown>, next: HttpHandlerFn):
    Observable<HttpEvent<unknown>> {
  return next(req);
}
```

As with functional router guards, functional interceptors have access to
Angular's DI via the `inject()` function synchronously within the
interceptor.

There are **major differences** in the registration of functional
interceptors as compared to the previous, class-based interceptors:

1. How they're provided:

Class-based interceptors are provided via the `HTTP_INTERCEPTORS` token
   using multi-providers.

Functional interceptors are provided with the `provideHttpInterceptors`
   function.

2. Scope of the interceptors:

Class-based interceptors must be provided in the same injector as
`HttpClient`. Each time `HttpClient` is provided (via `HttpClientModule`)
in an injector, it resets the list of interceptors. Therefore, it is not
possible to add new lazy-loaded interceptors without significant plumbing.

Functional interceptors are always registered to the current `HttpClient`,
even if provided in a child injector. In other words, functional
interceptors can be _added_ to an existing interceptor stack, and even
lazily loaded.

3. Deduplication of interceptors:

Class-based interceptors are never deduplicated. Multiple provisions of
`HTTP_INTERCEPTORS` with the same interceptor implementation will result in
that interceptor being registered multiple times.

Functional interceptors are deduplicated based on the identity of the
interceptor function. Providing the same interceptor multiple times, even in
different injectors, will not result in multiple registrations. Only the
first registration will count.

The first registration of a functional interceptor will determine the DI
context in which it runs. Subsequent registrations are no-ops, even if they
have different DI contexts.

In this way, functional interceptors solve some of the top problems and pain
points surrounding HTTP interceptors:

* Lazy loading of interceptors
* Multiple interceptor stacks depending on injection context
* Duplicate interceptors

Removing interceptors is not currently supported but is technically possible
under the design.

alxhub committed:
feat(http): `provide` functions as alternatives to NgModules

This commit adds several provider functions that operate as alternatives to
NgModule-based configuration of `HttpClient`:

* `provideHttpXsrfProtection` allows custom configuration of the XSRF
  protection functionality in `HttpClient`
* `provideHttpDisabledXsrfProtection` disables the built-in XSRF protection
* `provideHttpJsonpSupport` configures support for JSONP requests
* `provideHttpTesting` configures testing support for `HttpClient`

alxhub committed:
feat(http): make `HttpClient` available without importing its NgModule

This commit switches `HttpClient` and all of the associated machinery to be
`providedIn: 'root'`, allowing it to be injected and used without needing to
import `HttpClientModule`.

Importing `HttpClientModule` still overrides these providers, ensuring that
backwards compatibility is maintained with applications that only configure
`HttpClient` in non-root injection contexts.
